### PR TITLE
Test loading nitf from input stream and fix segfault

### PR DESCRIPTION
**Pull request by JonathanMeans**
@clydestanfield @mbojrab @asylvest #123

### CODE REVIEW

File: six/modules/c++/six.sicd/tests/test_load_from_input_stream.cpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        six::NITFReadControl reader;
+        reader.setXMLControlRegistry(&xmlRegistry);
+        reader.load(stream, schemaPaths);
+
+        streamComplexData = six::sicd::Utilities::getComplexData(reader);
+        six::sicd::Utilities::getWidebandData(reader, *(streamComplexData.get()), streamWidebandData);
+
+        // This tells the reader that it doesn't
+        // own an XMLControlRegistry
+        reader.setXMLControlRegistry(NULL);
```
Comment (asylvest) [inline six/modules/c++/six.sicd/tests/test_load_from_input_stream.cpp]: Why is this needed?  I don't see the reader trying to delete it anywhere...

File: six/modules/c++/six.sicd/tests/test_load_from_input_stream.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        xmlRegistry.addCreator(six::DataType::COMPLEX,
+            new six::XMLControlCreatorT<
+            six::sicd::ComplexXMLControl>());
+
+        six::NITFReadControl reader;
+        reader.setXMLControlRegistry(&xmlRegistry);
+        reader.load(stream, schemaPaths);
+
+        streamComplexData = six::sicd::Utilities::getComplexData(reader);
+        six::sicd::Utilities::getWidebandData(reader, *(streamComplexData.get()), streamWidebandData);
```
Comment (asylvest) [inline six/modules/c++/six.sicd/tests/test_load_from_input_stream.cpp]: Can just use `*streamComplexData`

File: six/modules/c++/six.sicd/tests/test_load_from_input_stream.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+        std::vector<std::string> schemaPaths;
+        if (argc == 3)
+        {
+            schemaPaths.push_back(argv[2]);
+        }
+        else
+        {
+            // In a normal installation, we can infer the path
+            const sys::Path progDirname =
```
Comment (asylvest) [inline six/modules/c++/six.sicd/tests/test_load_from_input_stream.cpp]: Can you make a utility function for this somewhere common that we can use for programs like this?

File: six/modules/c++/six/include/six/NITFReadControl.h [hunk 4]
```diff
@@ -222,6 +223,10 @@ class NITFReadControl : public ReadControl
 
         return (iCat == "LEG");
     }
+
+    // We need this for one of the load overloadings
+    // to prevent data from being deleted prematurely
+    mem::ScopedCopyablePtr<nitf::IOStreamReader> mStreamReader;
```
Comment (asylvest) [inline six/modules/c++/six/include/six/NITFReadControl.h]: So it's a tradeoff... this puts a weird special case as a member variable but in exchange doesn't change the public APIs for the `load(nitf::IOInterface& )` cases.  But what happens if we have another `CustomIO` or other overloading that also needs a smart pointer.  I think what would be cleaner is to have this simply hold a `mem::SharedPtr<nitf::IOInterface>` and always set it for all the overloadings.  That will change a couple load() overloadings, but I think it's worth it.  Thoughts?

**Comment by JonathanMeans** [comment]
I’m a little concerned that having that for every IOInterface would be somewhat misleading and make it harder to make clear what exactly we’re guarding against. But on the other hand, if there’s just a smart pointer for that special case, then someone tries to load the same reader with something else, then the stream from the old load would still be open, which seems like it could be bad. So I agree that a shared pointer to IOInterface is the way to go.

**Comment by JonathanMeans** [comment]
This is fixed.

File: six/modules/c++/six/source/Utilities.cpp [hunk 5]
```diff
@@ -1101,6 +1101,34 @@ std::auto_ptr<Data> six::parseDataFromString(const XMLControlRegistry& xmlReg,
...
 }
 
+std::string six::findSchemaPath(const std::string& progname)
+{
+    sys::OS os;
+    std::string currentDir = os.getCurrentExecutable(progname);
+
+    // Arbitrary depth to prevent infinite loop in case
+    // of weird project structure
+    const size_t MAX_DEPTH = 5;
```
Comment (asylvest) [inline six/modules/c++/six/source/Utilities.cpp]: This can be `static`

File: six/modules/c++/six.sicd/tests/test_load_from_input_stream.cpp [hunk 6]
```diff
@@ -0,0 +1,113 @@
+/* =========================================================================
+* This file is part of six-c++
```
Comment (asylvest) [inline six/modules/c++/six.sicd/tests/test_load_from_input_stream.cpp]: Should be `six.sicd-c++`

File: six/modules/c++/six.sicd/tests/test_load_from_input_stream.cpp [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+* =========================================================================
+*
+* (C) Copyright 2004 - 2016, MDA Information Systems LLC
+*
+* six-c++ is free software; you can redistribute it and/or modify
```
Comment (asylvest) [inline six/modules/c++/six.sicd/tests/test_load_from_input_stream.cpp]: Here too

File: six/modules/c++/six.sicd/tests/test_load_from_input_stream.cpp [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+        six::NITFReadControl reader;
+        reader.setXMLControlRegistry(&xmlRegistry);
+        reader.load(stream, schemaPaths);
+
+        streamComplexData = six::sicd::Utilities::getComplexData(reader);
+        six::sicd::Utilities::getWidebandData(reader, *streamComplexData,
+                streamWidebandData);
+
+         if (*streamComplexData != *fileComplexData)
```
Comment (asylvest) [inline six/modules/c++/six.sicd/tests/test_load_from_input_stream.cpp]: One extra space here

**Comment by asylvest** [comment]
This looks good.  Let me know when those couple trivial things are fixed and I'll merge.

**Comment by JonathanMeans** [comment]
They’re fixed.
